crates/task-impls/src/transactions.rs

```diff
@@ -235,68 +130,79 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
     }
 
     #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Transaction Handling Task", level = "error")]
-    async fn wait_for_transactions(&self) -> Option<Vec<TYPES::Transaction>> {
+    async fn wait_for_block(
+        &self,
+    ) -> Option<(AvailableBlockData<TYPES>, AvailableBlockHeaderInput<TYPES>)> {
         let task_start_time = Instant::now();
 
-        // TODO (Keyao) Investigate the use of transaction hash
-        // <https://github.com/EspressoSystems/HotShot/issues/1811>
-        // let parent_leaf = self.parent_leaf().await?;
-        // let previous_used_txns = match parent_leaf.tarnsaction_commitments {
-        //     Some(txns) => txns,
-        //     None => HashSet::new(),
-        // };
+        let last_leaf = self.consensus.read().await.get_decided_leaf();
+        let mut latest_block: Option<(
+            AvailableBlockData<TYPES>,
+            AvailableBlockHeaderInput<TYPES>,
+        )> = None;
+        while task_start_time.elapsed() < self.api.propose_max_round_time()
+            && latest_block.as_ref().map_or(true, |(data, _)| {
+                data.block_payload.get_transactions(&data.metadata).len()
+                    < self.api.min_transactions()
+            })
+        {
+            let mut available_blocks = match self
+                .builder_client
+                .get_available_blocks(last_leaf.block_header.payload_commitment())
+                .await
+            {
+                Ok(blocks) => blocks,
+                Err(err) => {
+                    error!(%err, "Couldn't get available blocks");
+                    continue;
+                }
+            };
+
+            available_blocks.sort_by_key(|block_info| block_info.block_size);
 
-        let receiver = self.transactions.subscribe().await;
+            let Some(block_info) = available_blocks.pop() else {
+                continue;
+            };
 
-        loop {
-            let all_txns = self.transactions.cloned().await;
-            debug!("Size of transactions: {}", all_txns.len());
-            // TODO (Keyao) Investigate the use of transaction hash
-            // <https://github.com/EspressoSystems/HotShot/issues/1811>
-            // let unclaimed_txns: Vec<_> = all_txns
-            //     .iter()
-            //     .filter(|(txn_hash, _txn)| !previous_used_txns.contains(txn_hash))
-            //     .collect();
-            let unclaimed_txns = all_txns;
+            // Don't try to re-claim the same block if builder advertises it again
+            if latest_block.as_ref().map_or(false, |block| {
+                block.0.block_payload.builder_commitment(&block.0.metadata) == block_info.block_hash
+            }) {
+                continue;
+            }
```

Comment:
If the builder is giving us multiple options for blocks,  and the first one matches the `latest_blocks` shouldn't we just try next block instead of starting the loop over and querying the builder again?

Reply:
The one we're checking is the "best" one, so if it's the same, I don't see much need to try other ones.
Maybe in the future when we have lists of transactions we want to include or more complex metrics by which we choose blocks, but for now any other block we check will be "worse" than the `latest_block`.

Reply:
I see I slightly misread the code, we are looping until the we find a block that is sufficient or we run out of time.

Reply:
Yes, that's correct